src/server/bitops_family.cc

```diff
@@ -183,60 +287,203 @@ bool SetBitValue(uint32_t offset, bool bit_value, std::string* entry) {
+    }
+  }
+};
+
+// =============================================
+// Set a new value to a given bit
+class BitNewValue {
   EngineShard* shard_ = nullptr;
   DbIndex index_ = 0;
 
  public:
-  explicit OverrideValue(const OpArgs& args) : shard_{args.shard}, index_{args.db_ind} {
+  explicit BitNewValue(const OpArgs& args) : shard_{args.shard}, index_{args.db_ind} {
   }
 
   OpResult<bool> Set(std::string_view key, uint32_t offset, bool bit_value);
 };
 
-OpResult<bool> OverrideValue::Set(std::string_view key, uint32_t offset, bool bit_value) {
+OpResult<bool> BitNewValue::Set(std::string_view key, uint32_t offset, bool bit_value) {
   auto& db_slice = shard_->db_slice();
-  DbIndex index = index_;
-
   DCHECK(db_slice.IsDbValid(index_));
+  bool old_value = false;
 
-  std::pair<PrimeIterator, bool> add_res;
-  try {
-    add_res = db_slice.AddOrFind(index_, key);
-  } catch (const std::bad_alloc&) {
-    return OpStatus::OUT_OF_MEMORY;
+  ElementAccess element_access{key, index_};
+  auto find_res = element_access.Find(shard_);
+
+  if (find_res != OpStatus::OK) {
+    return find_res;
   }
-  bool old_value = false;
-  PrimeIterator& it = add_res.first;
-  bool added = add_res.second;
-  auto UpdateBitMapValue = [&](std::string_view value) {
-    db_slice.PreUpdate(index, it);
-    it->second.SetString(value);
-    db_slice.PostUpdate(index, it, key, !added);
-  };
 
-  if (added) {  // this is a new entry in the "table"
+  if (element_access.IsNewEntry()) {
     std::string new_entry(GetByteIndex(offset) + 1, 0);
     old_value = SetBitValue(offset, bit_value, &new_entry);
-    UpdateBitMapValue(new_entry);
+    element_access.Commit(new_entry);
   } else {
-    if (it->second.ObjType() != OBJ_STRING) {
-      return OpStatus::WRONG_TYPE;
-    }
     bool reset = false;
-    std::string existing_entry{GetString(shard_, it->second)};
+    std::string existing_entry{element_access.Value()};
     if ((existing_entry.size() * OFFSET_FACTOR) <= offset) {  // need to resize first
       existing_entry.resize(GetByteIndex(offset) + 1, 0);
       reset = true;
     }
     old_value = SetBitValue(offset, bit_value, &existing_entry);
     if (reset || old_value != bit_value) {  // we made a "real" change to the entry, save it
-      UpdateBitMapValue(existing_entry);
+      element_access.Commit(existing_entry);
     }
   }
   return old_value;
 }
 
+// ---------------------------------------------------------
+
+std::string RunBitOperationOnValues(std::string_view op, BitOpValues values) {
+  // This function accept an operation (either OR, XOR, NOT or OR), and run bit operation
+  // on all the values we got from the database. Note that in case that one of the values
+  // is shorter than the other it would return a 0 and the operation would continue
+  // until we ran the longest value. The function will return the resulting new value
+  std::size_t max_len = 0;
+  std::size_t max_len_index = 0;
+
+  const auto BitOperation = [&]() {
```

Comment:
1. I do not think this function should be lambda. You can just pass maxlen other arguments to it and it can be external.
2. There is code duplication there - you return from each branch in the same way. This is caused by how you defined a c'tor and Run method of `BitOpString` class. In addition you pass to `Run()` - `std::string(max_len, 0)` string as an argument. why? if it's always the same, why not create it inside Run.  I would create a constructor that  accepts `values` and `max_len`, and Run function that accepts BitOperationFunc and SkipOperationFunc. This way you could define `operation`  before you do the branching and reduce the code duplication